### ISSUE TITLE
Authentication issues on it-hadoop-client

### ISSUE THREAD

**Issue by nsmith-** [dmwm/CMSSpark#24]
@dciangot and I have noticed the error:
`py4j.protocol.Py4JJavaError: An error occurred while calling None.org.apache.spark.api.java.JavaSparkContext.
: org.apache.hadoop.yarn.exceptions.YarnException: Failed to submit application_1548256949044_18633 to YARN : Failed to renew token: Kind: HDFS_DELEGATION_TOKEN, Service: ha-hdfs:analytix, Ident: (token for ncsmith: HDFS_DELEGATION_TOKEN owner=ncsmith@CERN.CH, renewer=nobody, realUser=, issueDate=1550609491439, maxDate=1551214291439, sequenceNumber=4836514, masterKeyId=2235)`
We each found a different workaround: one seems to be to not do `source hadoop-setconf.sh` as requested in https://hadoop-user-guide.web.cern.ch/hadoop-user-guide/getstart/client_edge_machine.html (and also remove it from https://github.com/dmwm/CMSSpark/blob/master/setup_lxplus.sh#L2-L3 ) (since it is done in https://github.com/dmwm/CMSSpark/blob/master/bin/run_spark#L101-L104 and I guess cannot be done twice?)
The other workaround seems to be to leave that in, but remove https://github.com/dmwm/CMSSpark/blob/master/bin/run_spark#L21-L23

I guess the question is, how can we patch this up to make it 'just work' ?

**Comment by vkuznet** [comment]
Before making any suggestion on how, please provide full description/steps which
lead to the problem. From login to somewhere, setting up the environment to
running specific task which lead to the problem.

We have several use-cases which we need to support, including:
- custom nodes
- lxplus setup
- edge machines
etc.

And, we need to maintain the script in a such a way that it should work in all
of them.

**Comment by nsmith-** [comment]
Yeah I gathered that from the length of `run_spark`.  I let @dciangot comment if he wants on his exact sequence.  My setup is:
```sh
ssh it-hadoop-client.cern.ch
source hadoop-setconf.sh analytix
cd CMSSpark/
source setup_lxplus.sh
cd src/python/CMSSpark
run_spark test.py --yarn
```
where the scripts have the following modification, to prevent the above error:
```diff
diff --git a/bin/run_spark b/bin/run_spark
index c731050..58c26a1 100755
--- a/bin/run_spark
+++ b/bin/run_spark
@@ -18,9 +18,9 @@ cmsspark=$droot/$1
 # enable simple secret to run in non-yarn mode
 conf=""
 hostname=`hostname -s`
-if [ "$hostname" != "vocms092" ]; then
-conf="--conf spark.authenticate.secret=cmsspark --conf spark.yarn.security.tokens.hive.enabled=false --conf spark.driver.port=5001 --conf spark.blockManager.port=5101 --conf spark.ui.port=520
-fi
+# if [ "$hostname" != "vocms092" ]; then
+#   conf="--conf spark.authenticate.secret=cmsspark --conf spark.yarn.security.tokens.hive.enabled=false --conf spark.driver.port=5001 --conf spark.blockManager.port=5101 --conf spark.ui.port
+# fi

 # look if we requested to show full log output, to disable spark output
 # client should setup his/her own log4j.properties via WMA_LOG4J environment variable
diff --git a/setup_lxplus.sh b/setup_lxplus.sh
index 00a3a73..af9e55c 100755
--- a/setup_lxplus.sh
+++ b/setup_lxplus.sh
@@ -1,5 +1,11 @@
 #!/bin/bash
-source /cvmfs/sft.cern.ch/lcg/views/LCG_93/x86_64-centos7-gcc62-opt/setup.sh
-source /cvmfs/sft.cern.ch/lcg/etc/hadoop-confext/hadoop-setconf.sh analytix
+# the it-hadoop-client edge nodes have a modified version
+# c.f. https://hadoop-user-guide.web.cern.ch/hadoop-user-guide/getstart/client_edge_machine.html
+if hash hadoop-setconf.sh; then
+  source hadoop-setconf.sh analytix
+else
+  source /cvmfs/sft.cern.ch/lcg/views/LCG_93/x86_64-centos7-gcc62-opt/setup.sh
+  source /cvmfs/sft.cern.ch/lcg/etc/hadoop-confext/hadoop-setconf.sh analytix
+fi
 export PATH=$PWD/bin:$PATH
 export PYTHONPATH=$PWD/src/python:$PYTHONPATH
```
```

**Comment by nsmith-** [comment]
Also, isn't the lxplus usage being deprecated by CERN IT?

**Comment by vkuznet** [comment]
No, I think quite the opposite. Many users don't have special account and
lxplus serve as common platform.

**Comment by vkuznet** [comment]
Nick,
for completeness, could you please post test.py as well that I can test your
procedure in different environments.
Thanks,
Valentin.

**Comment by vkuznet** [comment]
Nick,
turns out it is issue with your script rather than run_spark.

If I modify your script as following:

```
from pyspark import SparkConf, SparkContext
from pyspark.sql import SparkSession

ctx = SparkContext(appName='test')
spark = SparkSession(ctx)

#conf = SparkConf().setMaster("yarn").setAppName("CMS Working Set")
#sc = SparkContext(conf=conf)
#spark = SparkSession(sc)

readavro = spark.read.format("com.databricks.spark.avro")
fwjr = readavro.load("/cms/wmarchive/avro/fwjr/2019/01/19/*.avro")
print("done", type(fwjr))
print(fwjr.head())
```

it runs fine on lxplus. The cause of token problem came from the fact that you
setup your OWN configuration for spark at this line:

```
conf = SparkConf().setMaster("yarn").setAppName("CMS Working Set")
```

and as such the conf options passed by run_spark script were ignored (including
setting token).

I tweaked a little bit bin/run_spark to use local file as yours (code uploaded
to github), and I can run it without problem on lxplus:

```
bin/run_spark test.py
```

Bottom line, Spark setup is not trivial and we're trying to hide it from
end-users and possibly avoid issues. But if you write your own code you
on your own.

Said that I think the only change I need to apply is to use different
scripts for lxplus and it-hadoop-client nodes in setup_lxplus.sh script.

**Comment by nsmith-** [comment]
Actually in that case my test script was more or less a copy of `dbs_*.py` with the `run()` part cleared out except for setting up the session.  But ok, clearly spark is very temperamental.

```python
def run(date, fout, yarn=None, verbose=None):
    """
    Main function to run pyspark job. It requires a schema file, an HDFS directory
    with data and optional script with mapper/reducer functions.
    """
    sc = spark_context('cms', yarn, verbose)
    spark = SparkSession(sc)

    sc.stop()
```

**Comment by vkuznet** [comment]
Nick, I just tested this code:
```
from pyspark import SparkConf, SparkContext
from pyspark.sql import SparkSession

from pyspark.sql import SQLContext

ctx = SparkContext(appName='test')
sqlContext = SQLContext(ctx)
spark = SparkSession(ctx)

#conf = SparkConf().setMaster("yarn").setAppName("CMS Working Set")
#sc = SparkContext(conf=conf)
#spark = SparkSession(sc)

readavro = spark.read.format("com.databricks.spark.avro")
fwjr = readavro.load("/cms/wmarchive/avro/fwjr/2019/01/19/*.avro")
print("done", type(fwjr))
print(fwjr.head())
```
and it works with `run_spark` without changes you made. It works on lxplus or it-hadoop-client nodes. Therefore I don't see that I need to change anything in `run_spark` or `setup_lxplus.sh` scripts. The actual problem was that you defined your own configuration in your test.py which cause issue with token which you didn't specified in your configuration.

I'm closing this ticket. But if you think that things still does not work, please provide exact test.py you used, along with setup of your environment and run procedure that I can replicate.
